### PR TITLE
Fix Macro.to_string/2 for "unusual" atom remote calls

### PR DESCRIPTION
**Pull request by whatyouhide**
Aimed at fixing #5605. The reasoning is explained in the big comment in the code :)

### CODE REVIEW

File: lib/elixir/test/elixir/macro_test.exs [hunk 1]
```diff
@@ -271,6 +271,12 @@ defmodule MacroTest do
     assert Macro.to_string(quote do: :foo.(1, 2, 3)) == ":foo.(1, 2, 3)"
   end
 
+  test "unusual remote atom fun call to string" do
```
Comment (whatyouhide) [inline lib/elixir/test/elixir/macro_test.exs]: How do I call this test? halp

File: lib/elixir/lib/macro.ex [hunk 2]
```diff
@@ -841,7 +841,7 @@ defmodule Macro do
   defp call_to_string({:., _, [arg]}, fun),
     do: module_to_string(arg, fun) <> "."
   defp call_to_string({:., _, [left, right]}, fun),
-    do: module_to_string(left, fun) <> "." <> call_to_string(right, fun)
+    do: module_to_string(left, fun) <> "." <> if(is_atom(right), do: call_to_string_for_atom(right), else: call_to_string(right, fun))
```
Comment (michalmuskala) [inline lib/elixir/lib/macro.ex]: could this `is_atom` check be done in the function head and split those two cases into two heads? I think that would be cleaner than this nested if.

Comment (whatyouhide) [inline lib/elixir/lib/macro.ex]: Good call!

**Comment by josevalim** [comment]
I would like to not couple those two. I believe there are other places we do this coupling in Elixir and we should stop doing it. The reason is because in Elixir v1.5 we will likely support Unicode atoms, such as `:こんにちは世界`. This means the behaviour between those entries will be quite different. The logic that does such checking is in Inspect.Atom, we need to move it elsewhere. I bet we do similar mistakes elsewhere.

**Comment by whatyouhide** [comment]
@josevalim yep, I basically am taking advantage of exactly that logic in `Inspect.Atom`. Should we do this extraction before or after this PR? Also, should we move this logic into `Atom` directly? Should it be internal or public API?

**Comment by josevalim** [comment]
Let's use this PR as an opportunity to clean it up. About the place, I am not sure yet. It depends on what it receives and what it returns. It may as well belong to Macro.

**Comment by whatyouhide** [comment]
@josevalim few options:

```elixir
def Macro.escape_atom(:"foo bar") #=> "\"foo bar\""
def Atom.escape(:"foo bar") #=> "\"foo bar\""
```

Then `Atom.escape(false) == "false"`, `Atom.escape(:%{}) == "%{}" and so on.

I am not sure we should print `:Bar` as `:Bar` anymore though, maybe it's more consistent to have `:"Bar"`, otherwise we should still special case the first letter right?

**Comment by josevalim** [comment]
@whatyouhide I was thinking rather a function that receives an atom and classify it as `:identifier`, `:alias`, `:atom` or something similar. I think we need to decouple introspection of the data from the escaping/representation.

Also, it doesn't make sense to escape `:Bar` to `:"Bar"` given my earlier comment that we will allow other UTF-8 characters in the atom without quotes but we won't allow those without quotes in identifiers for now. So one way or the other, non-quoted atoms are a superset of non-quoted identifiers.

**Comment by whatyouhide** [comment]
@josevalim ah, that sounds like a good idea. Me and @lexmag discussed this yesterday and maybe a good place to put this function for now (public, but `@doc false`) could be `Inspect.Atom`. Otherwise, I think `Macro` is where it belongs (as we have things like `camelize` there, this "feels" in the same category). Wdyt?

**Comment by Kabie** [comment]
I asked about this a few days ago: http://stackoverflow.com/q/41072154/260985

At first, I thought that is something the compiler want to know, so I search for such a function in all the `:elixir_*` modules but didn't find any. Does the compiler not need such information?

**Comment by whatyouhide** [comment]
Closing in favor of #5629.

**Pull request closed without merging**